### ISSUE TITLE
<streambuf> : std::basic_streambuf doesn't support 64 bit properly

### ISSUE THREAD

**Issue by gilwohl** [microsoft/STL#388]
std::basic_streambuf doesn't support 64 bit properly, because it holds the counters in 32 bit ints:
	int _Gcount;	// length of read buffer
	int _Pcount;	// length of write buffer
	int *_IGcount;	// pointer to length of read buffer
	int *_IPcount;	// pointer to length of write buffer

Then e.g. in std::basic_streambuf::setg(...) you have:
        *_IGcount = static_cast<int>(_Last - _Next);
and forcing a cast to int prevents from using with a memory range bugger than ~2GB (even on Win64 builds).

**Comment by JeanPhilippeKernel** [comment]
@StephanTLavavej  can you assign me this one ?

**Comment by AdamBucior** [comment]
@JeanPhilippeKernel vNext branch is not available yet and as @StephanTLavavej said in #290 it probably won't be before the second half of calendar year 2020.

**Comment by JeanPhilippeKernel** [comment]
ooh 😥😥😪 sad news !

**Comment by StephanTLavavej** [comment]
Yeah - we're eager to resume work on vNext but first we have to finish migrating our build/test infrastructure, and then we need to manually port the changes we accumulated for vNext in a Team Foundation Version Control branch years ago (the codebase has evolved considerably since then, so the port will be time-consuming). Thanks for your patience, we hope vNext will be worth the wait :-)

**Comment by JeanPhilippeKernel** [comment]
Oh okay, Sure;) !

**Comment by cschreib-ibex** [comment]
I have just found out our library is affected by exactly that same issue. Do you know of a workaround we can use to bypass this limitation?

**Comment by StephanTLavavej** [comment]
I'm not familiar enough with iostreams to say whether such a workaround is reliable (despite working on this codebase for 13 years). @BillyONeal may be able to comment, as he's delved into iostreams' arcane mysteries.

**Comment by BillyONeal** [comment]
> I seem to have gone around it by changing our custom streambuffer class to always use a read range of at most `std::numeric_limits<int>::max()` bytes, and override `underflow()` to move this range further whenever more bytes need to be read. Feels a bit hackish though.

To the best of my understanding that should work (it's similar to how the stream buffers exposed don't contain all a file's contents) but we can't do it to the built in ones without breaking ABI.

**Comment by cschreib-ibex** [comment]
Thank you, we'll go for that workaround then. It passed all our tests, but of course they can't cover every possible case.

**Comment by JoelLinn** [comment]
Found this out the hard way…
I ended up overwriting `xsputn` and using my own pointer and length variables when inheriting from streambuf. It's enough for my use-case (ostream->pre-allocated memory buffer)

Please make sure this gets fixed in the next ABI version as 2GB isn‘t much these days.

**Comment by jimver04** [comment]
Any update on this ? see also my comment at : https://github.com/microsoft/STL/issues/578#issuecomment-1849502832
